### PR TITLE
Register fields for reflection in kubernetes-client

### LINKED ISSUE #39934: Kubernetes Client fails to create a new Pod in native executable because `Pod Overhead set without corresponding RuntimeClass defined Overhead.`
### Describe the bug

I have an application that creates a Pod in Kubernetes cluster. The Pod is very simple, contains only a volume with mounted ConfigMap and a single container. It works fine when I run it with `mvn quarkus:dev`. But fails after native compilation with the following error:

```
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://192.168.1.86:6443/api/v1/namespaces/myproject/pods. Message: pods "kekspose" is forbidden: pod rejected: Pod Overhead set without corresponding RuntimeClass defined Overhead. Received status: Status(apiVersion=v1, code=403, details=StatusDetails(causes=[], group=null, kind=pods, name=kekspose, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=pods "kekspose" is forbidden: pod rejected: Pod Overhead set without corresponding RuntimeClass defined Overhead, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Forbidden, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:507)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:524)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleCreate(OperationSupport.java:340)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handleCreate(BaseOperation.java:753)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handleCreate(BaseOperation.java:97)
	at io.fabric8.kubernetes.client.dsl.internal.CreateOnlyResourceOperation.create(CreateOnlyResourceOperation.java:42)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.create(BaseOperation.java:1148)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.create(BaseOperation.java:97)
	at cz.scholz.kekspose.Proxy.deployProxy(Proxy.java:81)
	at cz.scholz.kekspose.Kekspose.run(Kekspose.java:67)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2026)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at io.quarkus.picocli.runtime.PicocliRunner$EventExecutionStrategy.execute(PicocliRunner.java:26)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at io.quarkus.picocli.runtime.PicocliRunner.run(PicocliRunner.java:40)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:132)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
	at java.base@21.0.2/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://192.168.1.86:6443/api/v1/namespaces/myproject/pods. Message: pods "kekspose" is forbidden: pod rejected: Pod Overhead set without corresponding RuntimeClass defined Overhead. Received status: Status(apiVersion=v1, code=403, details=StatusDetails(causes=[], group=null, kind=pods, name=kekspose, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=pods "kekspose" is forbidden: pod rejected: Pod Overhead set without corresponding RuntimeClass defined Overhead, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Forbidden, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:660)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:640)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:589)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:549)
	at java.base@21.0.2/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	at java.base@21.0.2/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base@21.0.2/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.http.StandardHttpClient.lambda$completeOrCancel$10(StandardHttpClient.java:143)
	at java.base@21.0.2/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base@21.0.2/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base@21.0.2/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base@21.0.2/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.http.ByteArrayBodyHandler.onBodyDone(ByteArrayBodyHandler.java:52)
	at java.base@21.0.2/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base@21.0.2/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base@21.0.2/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base@21.0.2/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.vertx.VertxHttpRequest.lambda$null$1(VertxHttpRequest.java:122)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:279)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:261)
	at io.vertx.core.http.impl.HttpEventHandler.handleEnd(HttpEventHandler.java:76)
	at io.vertx.core.http.impl.HttpClientResponseImpl.handleEnd(HttpClientResponseImpl.java:250)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.lambda$new$0(Http1xClientConnection.java:421)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:255)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:134)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleEnd(Http1xClientConnection.java:709)
	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:313)
	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:293)
	at io.vertx.core.http.impl.Http1xClientConnection.handleResponseEnd(Http1xClientConnection.java:940)
	at io.vertx.core.http.impl.Http1xClientConnection.handleHttpMessage(Http1xClientConnection.java:810)
	at io.vertx.core.http.impl.Http1xClientConnection.handleMessage(Http1xClientConnection.java:774)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:159)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1475)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1338)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1387)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:530)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:469)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@21.0.2/java.lang.Thread.runWith(Thread.java:1596)
	at java.base@21.0.2/java.lang.Thread.run(Thread.java:1583)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:833)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:211)
```

This seems to be caused because the native executable sets the `.spec.overhead` fiels to an ampty object (`{}`) instead of keeping it `null`. When I explicitely set it to `null`:

```java
            new PodBuilder()
                // ...
                .withNewSpec()
                    // ...
                    .withOverhead(null)
                .endSpec()
                .build();
```

It seems to work fine.

### Expected behavior

The `.spec.overhead` field is set to `null` even without it being done in the `PodBuilder` and the application is able to create the new Pod even when compiled to a native executable.

### Actual behavior

A native executable fails with the error described above.

### How to Reproduce?

1. Write an application that creates a simple Pod object and creates it in Kubernetes cluster with `client.pods().inNamespace(namespace).resource(new PodBuilder()...build()).create();`
2. See that it works withotu native compilation
3. Compile it into native executable and observe that you get the error

The code where I run into this error can be found here: https://github.com/scholzj/kekspose/blob/main/src/main/java/cz/scholz/kekspose/Proxy.java#L81

### Output of `uname -a` or `ver`

Darwin Jakubs-MBP 23.4.0 Darwin Kernel Version 23.4.0: Fri Mar 15 00:10:42 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk version "21.0.2" 2024-01-16 OpenJDK Runtime Environment GraalVM CE 21.0.2+13.1 (build 21.0.2+13-jvmci-23.1-b30) OpenJDK 64-Bit Server VM GraalVM CE 21.0.2+13.1 (build 21.0.2+13-jvmci-23.1-b30, mixed mode, sharing)

### Mandrel or GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.9.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63)

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by jorsol**
- Fixes #39934

### CODE REVIEW

**Comment by geoand** [comment]
I can't say I understand this, can you explain a little?

Thanks

**Comment by jorsol** [comment]
> I can't say I understand this, can you explain a little?

Sure, long story short: Jackson uses reflection to search fields https://github.com/quarkusio/quarkus/issues/39934#issuecomment-2231104122:
https://github.com/FasterXML/jackson-databind/blob/jackson-databind-2.17.2/src/main/java/com/fasterxml/jackson/databind/introspect/AnnotatedFieldCollector.java
```java
private Map<String,FieldBuilder> _findFields(...) {
    for (Field f : cls.getDeclaredFields()) { // <- reflection alert!
...
```

Kubernetes Client model has been annotated with `JsonInclude` on the fields but not on the methods triggering issue https://github.com/quarkusio/quarkus/issues/39934

This PR registers all Kubernetes resources methods and fields for reflection potentially fixing the root issue.

**Comment by geoand** [comment]
Thanks.

I'll leave this to @manusa and @metacosm to review

**Comment by manusa** [comment]
> This PR registers all Kubernetes resources methods and fields for reflection potentially fixing the root issue.

I'm not sure what the initial reasons were to restrict the model registration to only a given set and to exclude fields.

This seems like a good temporary solution, but still I think that what @galderz [said originally](https://github.com/quarkusio/quarkus/issues/39934#issuecomment-2231104122) would make more sense quarkus-wise:

> The proper fix IMO would be to stop using reflection for discovering annotations. This would reduce binary executables, make native image builds faster and you would avoid issues like this because the jandex index has everything.

Related issues/PRs these changes affect:
- #27469 
- #27563

**Comment by jorsol** [comment]
> The proper fix IMO would be to stop using reflection for discovering annotations. This would reduce binary executables, make native image builds faster and you would avoid issues like this because the jandex index has everything.

Yes, that would be the proper way to fix it, but the thing is that is not Quarkus who uses reflection, is Jackson from here:
https://github.com/FasterXML/jackson-databind/blob/091d968751ef00150d22a788fe7f45b7cdcb337a/src/main/java/com/fasterxml/jackson/databind/introspect/AnnotatedFieldCollector.java#L73

So, while the correct approach would be to use Jandex, that probably should be done on `jackson-databind` itself.

**Comment by gsmet** [comment]
There has been a lot of discussion about the Kubernetes Client having a massive overhead when we were not careful about reflection registration so I think we need to be very careful about this (not saying it shouldn't go in but we need to measure the impact in terms of image size before we think about merging it).

**Comment by manusa** [comment]
> So, while the correct approach would be to use Jandex, that probably should be done on jackson-databind itself.

Agreed, hence my "This seems like a good temporary solution" statement :) I just wanted to keep on the record that the other task/issue should not be forgotten.

**Comment by jorsol** [comment]
> There has been a lot of discussion about the Kubernetes Client having a massive overhead when we were not careful about reflection registration so I think we need to be very careful about this (not saying it shouldn't go in but we need to measure the impact in terms of image size before we think about merging it).

3.13.0.CR1, final file size `169M`:
```
   29,093 reachable types   (86.3% of   33,710 total)
   45,578 reachable fields  (67.7% of   67,339 total)
  186,602 reachable methods (65.7% of  284,009 total)
   10,082 types, 2,974 fields, and 40,567 methods registered for reflection
       62 types,    67 fields, and    55 methods registered for JNI access
        4 native libraries: dl, pthread, rt, z
```

This patch, final file size `170M`:
```
   29,093 reachable types   (86.3% of   33,710 total)
   45,583 reachable fields  (67.7% of   67,342 total)
  186,603 reachable methods (65.7% of  284,008 total)
   10,082 types, 8,531 fields, and 40,567 methods registered for reflection
       62 types,    67 fields, and    55 methods registered for JNI access
        4 native libraries: dl, pthread, rt, z
```

There is a 1 MB overhead which is not bad at all. But please don't take my word for granted, anyone could test this to ensure there are no drawbacks.

**Comment by geoand** [comment]
At some point we really really have to look at the Kubernetes binary size issue  (due to Jackson) in a proper way.

cc @xstefank who is now going to working on the JOSDK (which is a heavy user of the client) and @mariofusco who has done a lot of work around Jackson

**Comment by gsmet** [comment]
>At some point we really really have to look at the Kubernetes binary size issue (due to Jackson) in a proper way.

Apart from being able to identify which areas of the client are used, I don't see a lot of options.
Because even if we end up generating code for JSON serialization, we would have to generate a massive amount of code if we have to support the whole client.

**Comment by geoand** [comment]
True as that may be, the fact that we are registering so much stuff for reflection AFAIU, is killing GraalVMs ability to do dead code elimination

**Comment by geoand** [comment]
We should also explore what more can be done on the client side to improve things

**Comment by mariofusco** [comment]
> @mariofusco who has done a lot of work around Jackson

I'm making good progresses with [my PoC](https://github.com/quarkusio/quarkus/pull/41063) trying to generate `ObjectMapper`s at build time via gizmo, thus having the advantage of avoiding any reflection and consequently also speeding up the marshalling process. I'm hoping to complete this task and have something ready for review in the next few days.

That said, I'm not sure if the work that I'm doing could be also applied to this specific issue, but my understanding is (please correct me if I'm wrong) that here we need to convert in json some specific pojos that are mapping the kubernetes configuration. If that's true couldn't we also develop some ad-hoc `ObjectMapper`s (if the pojos are known upfront we don't need any code generation) for those classes instead of relying on out-of-the-box reflection-based Jackson marshalling?

**Comment by geoand** [comment]
>  but my understanding is (please correct me if I'm wrong) that here we need to convert in json some specific pojos that are mapping the kubernetes configuration.

I'm not the SME (@manusa is), but this is my understanding as well

**Comment by manusa** [comment]
There are several places in the client where deserialization (json-2-Pojo) takes place:
- Reading HTTP responses from API client
- Loading resources from storage 
- Reading the Kubernetes configuration (`.kube/config` et al)

The [`KubernetesDeserializer`](https://github.com/fabric8io/kubernetes-client/blob/25ecf8460eb132ae2945342b33f64a4240a2b91c/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java) takes care of deserializing JSON nodes to the required target types (via reflection) as long as there's a model artifact that contains the desired classes.

However, long time ago we introduced the [`GenericKubernetesResource`](https://github.com/fabric8io/kubernetes-client/blob/25ecf8460eb132ae2945342b33f64a4240a2b91c/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/GenericKubernetesResource.java) which is the fallback type in case no specific class is registered or available to deserialize the JSON node.

This means that if you application is not doing any type-specific operation, it might just work by registering `GenericKubernetesResource` for reflection (+ those specific the `.kube/config`, etc.).
An example of this kind of application can be something that reads a bunch of YAMLs and deploys them to Kubernetes.

One of the problems is that the Kubernetes Client Quarkus extension is to broad in scope. Plus the `io.fabric8:kubernetes-client` is also too broad and includes transitive dependencies for all of the model modules.

I still have a pending task to create a blog post showcasing how to create Quarkus Kubernetes app with a limited set of models. But the thing is that users are the ones who know in advance what types their application will be needing.

One of the things I think might be hard to solve (but maybe I'm wrong and GraalVM/Quarkus can actually take care of this) is something like:

```java
KubernetesResource resource = kubernetesClient.resource($inputStream).item();
if (resource instanceof Service) {
  System.out.println("service");
} else {
  System.out.println("not-service");
}
```

In this case, if the `io.fabric8:kubernetes-model-core` artifact is available at runtime and `Service` is registered for reflection, then the output will be `service`. Otherwise, the stream will be deserialized as a `GenericKubernetesResource` and the output will be `not-service`.

**Comment by metacosm** [comment]
> > The proper fix IMO would be to stop using reflection for discovering annotations. This would reduce binary executables, make native image builds faster and you would avoid issues like this because the jandex index has everything.
> 
> Yes, that would be the proper way to fix it, but the thing is that is not Quarkus who uses reflection, is Jackson from here: https://github.com/FasterXML/jackson-databind/blob/091d968751ef00150d22a788fe7f45b7cdcb337a/src/main/java/com/fasterxml/jackson/databind/introspect/AnnotatedFieldCollector.java#L73
> 
> So, while the correct approach would be to use Jandex, that probably should be done on `jackson-databind` itself.

maybe it might be worth investigating whether we could have a Jandex-based collector in Jackson?

**Comment by metacosm** [comment]
> There are several places in the client where deserialization (json-2-Pojo) takes place:
> 
>     * Reading HTTP responses from API server
> 
>     * Loading resources from storage
> 
>     * Reading the Kubernetes configuration (`.kube/config` et al)
> 
> 
> The [`KubernetesDeserializer`](https://github.com/fabric8io/kubernetes-client/blob/25ecf8460eb132ae2945342b33f64a4240a2b91c/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java) takes care of deserializing JSON nodes to the required target types (via reflection) as long as there's a model artifact that contains the desired classes.
> 
> However, long time ago we introduced the [`GenericKubernetesResource`](https://github.com/fabric8io/kubernetes-client/blob/25ecf8460eb132ae2945342b33f64a4240a2b91c/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/GenericKubernetesResource.java) which is the fallback type in case no specific class is registered or available to deserialize the JSON node.
> 
> This means that if you application is not doing any type-specific operation, it might just work by registering `GenericKubernetesResource` for reflection (+ those specific the `.kube/config`, etc.). An example of this kind of application can be something that reads a bunch of YAMLs and deploys them to Kubernetes.
> 
> One of the problems is that the Kubernetes Client Quarkus extension is too broad in scope. Plus the `io.fabric8:kubernetes-client` is also too broad and includes transitive dependencies for all of the model modules.
> 
> I still have a pending task to create a blog post showcasing how to create Quarkus Kubernetes app with a limited set of models. But the thing is that users are the ones who know in advance what types their application will be needing.
> 
> One of the things I think might be hard to solve (but maybe I'm wrong and GraalVM/Quarkus can actually take care of this) is something like:
> 
> ```java
> KubernetesResource resource = kubernetesClient.resource($inputStream).item();
> if (resource instanceof Service) {
>   System.out.println("service");
> } else {
>   System.out.println("not-service");
> }
> ```
> 
> In this case, if the `io.fabric8:kubernetes-model-core` artifact is available at runtime and `Service` is registered for reflection, then the output will be `service`. Otherwise, the stream will be deserialized as a `GenericKubernetesResource` and the output will be `not-service`.

Wild idea (because this would change a lot of how the client work currently): could we keep the current model API but have the objects backed by a generic resource. Instead of all model objects being generated individually with appropriate fields, we would just getting the accessors, which would take care of accessing the appropriate key-value in the underlying map?

Maybe that would simplify the serialization aspects? That said, this might not help size in memory or performance (since you'd have to access a map instead of a direct field access)…

**Comment by jorsol** [comment]
> maybe it might be worth investigating whether we could have a Jandex-based collector in Jackson?

Jackson might need to allow both collectors, the Reflection-based and the Jandex-based collector, and be configurable since not everyone could use Jandex.

---

But, can we keep focus on the current issue here? I agree is important to improve and reduce reflection in most libraries used by Quarkus, but right now the issue might be solved by just registering the fields with an overhead of just 1MB in the final size of the binary.

**Comment by geoand** [comment]
> I agree is important to improve and reduce reflection in most libraries used by Quarkus

For us, the current situation is pretty bad, that's why we are having the discussion about potential improvements. If it doesn't happen now, it will slip yet again...

In any case, the PR itself makes sense

**Comment by manusa** [comment]
> Wild idea (because this would change a lot of how the client work currently): could we keep the current model API but have the objects backed by a generic resource. Instead of all model objects being generated individually with appropriate fields, we would just getting the accessors, which would take care of accessing the appropriate key-value in the underlying map?

Maybe we can discuss this during a call. This might get complicated especially when trying to deal with deserialization of non-standard fields, etc. Also hard for the sundrio builder-generation integration.

**Comment by mariofusco** [comment]
> > Wild idea (because this would change a lot of how the client work currently): could we keep the current model API but have the objects backed by a generic resource. Instead of all model objects being generated individually with appropriate fields, we would just getting the accessors, which would take care of accessing the appropriate key-value in the underlying map?
> 
> Maybe we can discuss this during a call. This might get complicated especially when trying to deal with deserialization of non-standard fields, etc. Also hard for the sundrio builder-generation integration.

I would love to also discuss this in a call, since it could even be related to the work that I'm doing for the PoC that I mentioned in one of my former comments.

**Comment by galderz** [comment]
> To clarify the current issue:
> 
> kubernetes-client model contains classes with Jackson annotations, and on previous versions of KC the fields had an additional annotation (`@JsonInclude`) that the methods don't:
> 
> ```java
>     @JsonProperty("labels")
>     @JsonInclude(JsonInclude.Include.NON_EMPTY)
>     private Map<String, String> labels = new LinkedHashMap<String, String>();
>     
>     @JsonProperty("labels")
>     public Map<String, String> getLabels() {
>         return labels;
>     }
> 
>     @JsonProperty("labels")
>     public void setLabels(Map<String, String> labels) {
>         this.labels = labels;
>     }
> ```
> 
> Since the fields were not registered, the `@JsonInclude` was not evaluated causing a different/wrong behavior in native-image.
> 
> If the reflection can be avoided that would be great, but for the time being, this change only registers correctly the fields of the KC model.

Note that there's an alternative solution here, which is that methods always have the same annotations as fields. Then you "just" have to register methods for reflection and you can ignore fields. But if there's functionality that relies on annotations that are only in fields then it would break.

File: extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java [hunk 1]
```diff
@@ -68,16 +63,15 @@ public class KubernetesClientProcessor {
...
-    private static final DotName KUBERNETES_RESOURCE = DotName.createSimple(KubernetesResource.class.getName());
-    private static final DotName VALIDATION_SCHEMA = DotName.createSimple(ValidationSchema.class.getName());
+    private static final DotName KUBERNETES_RESOURCE = DotName
+            .createSimple(KubernetesResource.class.getName());
     private static final DotName KUBERNETES_RESOURCE_LIST = DotName
             .createSimple(KubernetesResourceList.class.getName());
-    private static final DotName KUBE_SCHEMA = DotName.createSimple(KubeSchema.class.getName());
     private static final DotName VISITABLE_BUILDER = DotName.createSimple(VisitableBuilder.class.getName());
+    private static final DotName VALIDATION_SCHEMA = DotName.createSimple(ValidationSchema.class.getName());
+    private static final DotName KUBE_SCHEMA = DotName.createSimple(KubeSchema.class.getName());
```
Comment (zakkak) [inline extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java:72]: Nitpick: Are these changes applied automatically by formatting? Can you please avoid them?

**Comment by manusa** [comment]
> Note that there's an alternative solution here, which is that methods always have the same annotations as fields. Then you "just" have to register methods for reflection and you can ignore fields. But if there's functionality that relies on annotations that are only in fields then it would break.

This will be definitely fixed when [6.13.2](https://github.com/fabric8io/kubernetes-client/issues/6097) is released (probably on 2024-08-05).

However, users who generate their own types and don't annotate them uniformly might still run into the issue.

> I would love to also discuss this in a call, since it could even be related to the work that I'm doing for the PoC that I mentioned in one of my former comments.

I created a discussion on the Fabric8 Kubernetes Client repo to follow up on this and avoid adding more noise to the PR: https://github.com/fabric8io/kubernetes-client/discussions/6199

**Review by zakkak — APPROVED** [review]
I agree with what has been said so far. As far as I am concerned the PR looks OK as an immediate solution.

FWIW the size increase is indeed not huge, but still noticeable.

In `kubernetes-client` IT I see an increase of ~.5MB in the image size (or 0.3% of the total size and 0.8% of the heap size) which is the result of registering 10973 more fields for reflection.

In `openshift-client` IT I see a similar increase of ~.5MB in the image size (or 0.4% of the total size and 0.9% of the heap size) which is the result of registering 11635 more fields for reflection.

File: extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java (hunk 1 above)
Comment (jorsol) [reply extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java]: Done!

**Review by zakkak — APPROVED** [review]

File: extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java (hunk 1 above)
Comment (zakkak) [reply extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java]: Thanks!

**Comment by gsmet** [comment]
Yeah same here, given the current situation, we can pay the price for it.

**Pull request merged**